### PR TITLE
Fix indexDB putObject concurrency issue and cache.

### PR DESCRIPTION
**Pull request by yuchenshi**
WANT_LGTM=all

### CODE REVIEW

File: packages-exp/auth-exp/src/platform_browser/persistence/local_storage.ts [hunk 1]
```diff
@@ -209,7 +206,6 @@ class BrowserLocalPersistence
   }
 
   _addListener(key: string, listener: StorageEventListener): void {
-    this.localCache[key] = this.storage.getItem(key);
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/platform_browser/persistence/local_storage.ts:212]: i think you still need this, it warms up the cache from existing data when we register a listener. without this you could have a situation like:
- add value
- attach listener
- delete value
and the listener wouldn't fire

File: packages-exp/auth-exp/src/platform_browser/persistence/indexed_db.ts [hunk 2]
```diff
@@ -127,20 +127,11 @@ export async function _putObject(
   key: string,
   value: PersistenceValue | string
 ): Promise<void> {
-  const getRequest = getObjectStore(db, false).get(key);
-  const data = await new DBPromise<DBObject | null>(getRequest).toPromise();
-  if (data) {
```
Comment (avolkovi) [inline packages-exp/auth-exp/src/platform_browser/persistence/indexed_db.ts:132]: i think you still need this branch, my understanding is it's required for updating vs creating objects

File: packages-exp/auth-exp/src/platform_browser/persistence/local_storage.ts (hunk 1 above)
Comment (avolkovi) [reply packages-exp/auth-exp/src/platform_browser/persistence/local_storage.ts:212]: nvm, you'll have written to the cache regardless during the initial write

File: packages-exp/auth-exp/src/platform_browser/persistence/indexed_db.ts (hunk 2 above)
Comment (avolkovi) [reply packages-exp/auth-exp/src/platform_browser/persistence/indexed_db.ts:132]: discussed IRL

**Review by avolkovi — APPROVED** [review]

File: packages-exp/auth-exp/src/platform_browser/persistence/indexed_db.ts (hunk 2 above)
Comment (yuchenshi) [reply packages-exp/auth-exp/src/platform_browser/persistence/indexed_db.ts:132]: `The put() method of the IDBObjectStore interface updates a given record in a database, or inserts a new record if the given item does not already exist.` -- [MDN](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/put)

File: packages-exp/auth-exp/src/platform_browser/persistence/local_storage.ts (hunk 1 above)
Comment (yuchenshi) [reply packages-exp/auth-exp/src/platform_browser/persistence/local_storage.ts:212]: Cache is always updated on any data changes (add/put/remove) regardless of # listeners so there is no need for this step. It only complicates things.

**Review by samhorlbeck — APPROVED** [review]
Looks good pending tests passing

File: packages-exp/auth-exp/src/platform_browser/persistence/local_storage.ts (hunk 1 above)
Comment (yuchenshi) [reply packages-exp/auth-exp/src/platform_browser/persistence/local_storage.ts:212]: Actually, some proactive populating helps reduce spuriously triggering and fixes some of the tests. I've added a conditioned clause back and to indexedDb to stay consistent.

**Pull request merged**